portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -43,15 +46,25 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
     List<Comment> comments = new ArrayList<>();
     for (Entity entity : results.asIterable()) {
+      // Get comment's properties.
       long id = entity.getKey().getId();
-      String name = (String) entity.getProperty("name");  //NEW
+      String name = (String) entity.getProperty("name"); 
       String text = (String) entity.getProperty("text");
       long time = (long) entity.getProperty("time");
+      
+      // Pass user's comment text to LanguageServiceClient and receive sentiment score.
+      Document doc = Document.newBuilder().setContent(text).setType(Document.Type.PLAIN_TEXT).build();
+      LanguageServiceClient languageService = LanguageServiceClient.create();
+      Sentiment sentiment = languageService.analyzeSentiment(doc).getDocumentSentiment();
+      float score = sentiment.getScore();
+      languageService.close();
```

Comment:
Consider refactoring this out into its own method. 

Also, did you consider doing the analysis on POST instead of GET? Doing it on GET means you'll call the API every time you fetch all the comments. If someone loads your page 100 times, this means calling the Natural Language API 100 times for each comment 😄

Reply:
Oh yeah, that was actually the first thing I tried! When I did the analysis on POST, for some reason none of the comments loaded, and I'm not entirely sure why. 🤔

Reply:
It might be because you were trying to fetch a property that wasn't set (the sentiment score property). Try changing the code to set the sentiment score on `POST` instead, running `mvn appengine:devserver`, and running this command from the [cheatsheet](https://docs.google.com/document/d/1yeDuDfEoVh2RB2wC-KrhkXNlekc15RxkPy8p-G9U4tA/edit#heading=h.2zjm1coh4mri) in a separate terminal:
```sh
curl -v localhost:8080/data
```
It'll return HTTP 500 if an internal error occurs if you're trying to read the sentiment score but it doesn't exist for a record. You can fix that by either:
- Doing a manual backfill. You can go to the [Datastore Entities](https://console.cloud.google.com/datastore/entities/query/kind?project=tshafikah-sps-spring20) page and set the `score` properties for any records missing it.
- Doing an automatic backfill on the next `GET` request. In `doGet()`,
  1. Check first if the `score` property exists
  2. If it does, add the comment to the `comments` list as is.
  3. If it doesn't, calculate the score and write the comment back to Datastore before adding it to the `comments` list.

Reply:
I moved the analysis to be done in `POST`, and manually backfill score properties for all comments through the Datastore Entities page. I then added `float score = (float) entity.getProperty("score");` at line 63 to get each comment's score.

When I ran `mvn appengine:devserver` then `curl -v localhost:8080/data` on a separate terminal, it gave me the following error:
`<pre>    Server Error</pre></p><h3>Caused by:</h3><pre>java.lang.ClassCastException: java.lang.Double cannot be cast to java.lang.Float
        at com.google.sps.servlets.DataServlet.doGet(DataServlet.java:63)`

Could this be a data type error?

Reply:
Ah, yes it's a data type error. `float`s cannot be casted to `double`. You'll need to do `double score = (double) entity.getProperty("score");`

Reply:
It works, thank you!